### ISSUE TITLE
fuse: Add directory entries `.` and `..`

### ISSUE THREAD

**Issue by fd0** [restic/restic#802]
The restic fuse mount does not include the directory entries `.` and `..`, which are not necessary for browsing the snapshots via fuse (so `cd ..` works just fine), but this may be confusing for users.

This issue tracks adding those entries.

**Comment by jgfrm** [comment]
I will investigate this.

**Comment by fd0** [comment]
Ok, cool!

**Comment by jgfrm** [comment]
For the backup tree, it is working.
However, there is one problem remaining.
The generated directories (representing the snapshots) and the snapshots dir should have a sound inode that do not collide with the inodes for the files and dirs in the backup.

**Comment by fd0** [comment]
Hm, good point. Maybe use very small inode numbers? Or a mapping table that does translation in case there is a conflict?

**Comment by jgfrm** [comment]
Yes, a translation table was also my thought.
All in all it requires quite some work:
- mapping table to avoid inode collision
- the current code (mainly snapshot.go) collapses three things:
* serve as fuse root
* provide snapshots dir
* provide the dirs for the individual snapshots

So far, I separated the last two parts.
It will take some time ...

**Comment by fd0** [comment]
Sure, no rush. Let me know when you need feedback on sample code or so.

**Comment by jgfrm** [comment]
Well, after some thinking, I discovered the following:
Restic backups can come from multiple physical disks. So, the same inode can actually refere multiple, and different, files. This can happen if these files are on different disks.
In principle, this is represented in the meta data of the backup, namely in the inode and dev field.
Fuse (at least the bazil implementation) only supports the inode field for stat.
So, I guess there should be a generation of the fuse inod, based upon the inode/dev metadata in the backup. 
Problem: both the inode and dev fields in the backup meta data are uint64, whereas the inode field in fuse is also uint64. 

Any suggestions how to deal with this?

**Comment by fd0** [comment]
Hm, you're right. I think we need a global mapping table for this use case anyway, so we can assign a different inode when a collision occurs (e.g. `orig_inode XOR orig_device` or so)

**Comment by jgfrm** [comment]
The problem is even worse; multiple snapshots of the same disk (in time) have files with the same inode and device combination. They are however all accessable via the same fuse mount. Therefore the inodes are nout unique across the fuse file system.

The overall problem is that inodes (represented as an uint64) of different disks and/or snapshots should be mapped into one fuse inode space (also uint64). Obviously, we can end up in the situation that are more inodes than fit into one uint64 space, theoretically 2^64*#disks*#snapshots.

The practical side of the problem is that ext4 seems to have 32 bits for the inode.

So my proposal the following:
1 The 64 bits of the uint64 that is used to represent the inode is used as follows:
- bits 57-64 9 (snapshot bits) are used to represent the snapshot id (via a mapping table of the restic snapshot ID and the snapshot bits), the mount point, and the "snapshots" dir. This allows for 2^8-2 (254) snapshots accessible via fuse. This seems to me a reasonable amount.
- bits 50-57 (device bits) are used to represent the device id as stored in the metadata, again with a mapping table between the bits and the device id. This allows for max 256 disks in a restic backup.
- bit 49 is used to indicate the use of bits 1-48. 0 means inode based on derivation, 1 means inode based on mapping tables
- bit 1-48 i used to represent inodes as they come from the meta data. Since ext4 seems to have 32 bits for inodes, this fits easily. In case the inode is > 2^48, mapping tables are used. I think the need for mapping tables is very limited.
- If bit49 == 0 then the fuse inode is calculated as inode (bit1...48) | device-bits | snapshot bits
- if the inode of the metadata > 2^48, then it is mapped into a mapping table, which gives another 2^48 to deal with such exceptions.

The advantage of this solution is that in normal cases we can calculate the fuse inode by using the inode in the meta data, the device id, and the snapshot. This avoids time consuming table lookups. In case we run out of inode space there is a facility to deal with these situations (bit49=1).

What do you think?

**Comment by fd0** [comment]
Oh, that's indeed a bigger problem than I imagined. To be honest, I think your proposal is too complex. What about just having a lookup table for ALL inodes, e.g. like this:
```go
type inodeKey struct {
     Snapshot uint64
     Inode uint64
     Device uint64
}

var inodes = make(map[inodeKey]uint64)
```

Then compute the key as follows:
```go
key.Snapshot = crc64(snapshot.ID)
key.Inode = node.Inode
key.Device = node.Dev
```

And assign a new random inode and use that for the remainder of the fuse session:
```go
inodes[key] = rand.Uint64()
```

(access syncronisation is missing here, obviously)

**Comment by jgfrm** [comment]
I see two problems hete:
1. You should build a table for all files. In any serious filesystem that can become large, with all the performance problems that come with that.
The solution I propose calculates the inode on the fly, and requires only a table for exceptions, which should not occur for ext4 filesystems at all. 2. Randomly assigning has the (limited) danger of inode collisions

**Comment by fd0** [comment]
I've fixed this issue in #1048, I'm using a deterministic pseudo-random inode-calculator provided by the fuse library: http://godoc.org/bazil.org/fuse/fs#GenerateDynamicInode
